Fallback to proc when ebpf timestamps are wrong

schu opened the pull request:
cc @alban

iaguis commented on probe/endpoint/ebpf.go:
```diff
@@ -99,7 +100,11 @@ var lastTimestampV4 uint64
 	if lastTimestampV4 > e.Timestamp {
-		log.Errorf("ERROR: late event!\n")
+		// A kernel bug can cause the timestamps to be wrong, i.e. Ubuntu with Linux 4.4.0-47.68
+		// See https://github.com/iovisor/bcc/issues/790#issuecomment-263704235
```
You can also reference #2334, it has all the info condensed.

schu commented:
Works in a manual test like https://github.com/schu/scope/commit/6febb7dd7bf8fc510009e62ae93344ca0bee8ca9 (https://circleci.com/gh/schu/scope/308) For an integration test we would have to prepare and use a separate note with an older kernel w/o the fix. Not worth the effort IMHO.

2opremio commented on probe/endpoint/connection_tracker.go:
```diff
@@ -88,8 +88,19 @@ func (t *connectionTracker) ReportConnections(rpt *report.Report) {
+			t.ebpfTracker = nil
+		} else {
+			t.performEbpfTrack(rpt, hostNodeID)
+			return
```
check `!t.ebpfTracker.isDead()`, which allows to remove the `else` part since it returns

2opremio replied:
i.e. 

```
if t.ebpfTracker != nil {
  if !t.ebpfTracker.isDead() {
     t.performEbpfTrack(rpt, hostNodeID)
     return
  }
  log.Warnf("ebpf tracker died, gently falling back to proc scanning")
  // fallback
}

```

2opremio commented on probe/endpoint/ebpf.go:
```diff
@@ -99,7 +100,10 @@ var lastTimestampV4 uint64
 	if lastTimestampV4 > e.Timestamp {
-		log.Errorf("ERROR: late event!\n")
+		// See https://github.com/weaveworks/scope/issues/2334
+		log.Debugf("tcp tracer received event with timestamp %v even though the last timestamp was %v. Stopping the eBPF tracker.", e.Timestamp, lastTimestampV4)
```
I think this should be an error

2opremio replied:
Also, explaning the probable cause of this (the kernel bug) and indicating that upgrading the kernel will likely fix it would be helpful.

2opremio commented:
> For an integration test we would have to prepare and use a separate node with an older kernel w/o the fix. Not worth the effort IMHO.

You could mock the failure and add a unit-test

schu commented:
Updated.

> You could mock the failure and add a unit-test

Mocking the ebpf part would require too much work in my opinion (as we would have to mock most of it). I have added a simple test to see if we set the tracker to dead after receiving events out of order.

2opremio commented:
This didn't work as expected. This is what I got when testing it in WeaveCloud's dev environment (the ebpf error continues in a loop).

```
time="2017-03-21T12:20:45Z" level=info msg="publishing to: https://7xs181ap6kabbaz3ttozt37i3ebb5e4b@frontend.dev.weave.works, https://wwmogdoe75wfj3fjfniz6yhfm1h9zzm5@cloud.weave.works, scope.kube-system.svc.cluster.local:80" 
<probe> INFO: 2017/03/21 12:20:45.750334 command line args: --logtostderr=true --mode=probe --no-app=true --probe.conntrack.buffersize=2097152 --probe.docker=true --probe.docker.bridge=cbr0 --probe.ebpf.connections=true --probe.http.listen=:4041 --probe.kubernetes=tru
e --probe.resolver=10.0.0.10:53 --weave=false https://7xs181ap6kabbaz3ttozt37i3ebb5e4b@frontend.dev.weave.works https://wwmogdoe75wfj3fjfniz6yhfm1h9zzm5@cloud.weave.works scope.kube-system.svc.cluster.local:80
<probe> INFO: 2017/03/21 12:20:45.750461 probe starting, version 2a61656, ID 2a3f76c3ed315deb
<probe> INFO: 2017/03/21 12:20:45.829637 Control connection to frontend.dev.weave.works starting
<probe> INFO: 2017/03/21 12:20:45.866573 Control connection to cloud.weave.works starting
<probe> INFO: 2017/03/21 12:20:45.892792 Control connection to scope.kube-system.svc.cluster.local starting
<probe> INFO: 2017/03/21 12:20:47.819510 kubernetes: targeting api server https://10.0.0.1:443
<probe> ERRO: 2017/03/21 12:20:47.897545 tcp tracer received event with timestamp 9164623054914167 even though the last timestamp was 9164623056153635. Stopping the eBPF tracker.
<probe> ERRO: 2017/03/21 12:20:47.926527 tcp tracer received event with timestamp 9164623054136294 even though the last timestamp was 9164623055631102. Stopping the eBPF tracker.
<probe> ERRO: 2017/03/21 12:20:48.168299 plugins: problem loading: no such file or directory
<probe> INFO: 2017/03/21 12:20:48.170005 Profiling data being exported to :4041
<probe> INFO: 2017/03/21 12:20:48.170117 go tool proof http://:4041/debug/pprof/{profile,heap,block}
<probe> ERRO: 2017/03/21 12:20:48.171029 tcp tracer received event with timestamp 9164624052746315 even though the last timestamp was 9164624055753098. Stopping the eBPF tracker.
<probe> ERRO: 2017/03/21 12:20:48.205917 tcp tracer received event with timestamp 9164624052912341 even though the last timestamp was 9164624053316166. Stopping the eBPF tracker.
<probe> ERRO: 2017/03/21 12:20:48.337777 conntrack stderr:NOTICE: Netlink socket buffer size has been set to 4194304 bytes.
<probe> ERRO: 2017/03/21 12:20:48.652453 tcp tracer received event with timestamp 9164624054080901 even though the last timestamp was 9164624055925277. Stopping the eBPF tracker.
<probe> ERRO: 2017/03/21 12:20:49.094904 tcp tracer received event with timestamp 9164625053710135 even though the last timestamp was 9164625055879401. Stopping the eBPF tracker.
<probe> ERRO: 2017/03/21 12:20:49.111708 tcp tracer received event with timestamp 9164625053335726 even though the last timestamp was 9164625054386441. Stopping the eBPF tracker.
<probe> ERRO: 2017/03/21 12:20:49.378834 tcp tracer received event with timestamp 9164625052644795 even though the last timestamp was 9164625055487794. Stopping the eBPF tracker.
<probe> WARN: 2017/03/21 12:20:49.530015 ebpf tracker died, gently falling back to proc scanning
<probe> ERRO: 2017/03/21 12:20:49.554308 tcp tracer received event with timestamp 9164625056086104 even though the last timestamp was 9164625056308378. Stopping the eBPF tracker.
<probe> ERRO: 2017/03/21 12:20:49.578644 tcp tracer received event with timestamp 9164625052604987 even though the last timestamp was 9164625056236728. Stopping the eBPF tracker.
<probe> ERRO: 2017/03/21 12:20:49.598553 tcp tracer received event with timestamp 9164625052491960 even though the last timestamp was 9164625052604987. Stopping the eBPF tracker.
<probe> ERRO: 2017/03/21 12:20:49.752892 conntrack stderr:NOTICE: Netlink socket buffer size has been set to 4194304 bytes.
<probe> ERRO: 2017/03/21 12:20:50.341305 tcp tracer received event with timestamp 9164626052415625 even though the last timestamp was 9164626053456793. Stopping the eBPF tracker.
<probe> ERRO: 2017/03/21 12:20:50.456562 tcp tracer received event with timestamp 9164626053328268 even though the last timestamp was 9164626054166952. Stopping the eBPF tracker.
<probe> ERRO: 2017/03/21 12:20:50.495565 tcp tracer received event with timestamp 9164626053263961 even though the last timestamp was 9164626054711229. Stopping the eBPF tracker.
<probe> ERRO: 2017/03/21 12:20:50.539413 tcp tracer received event with timestamp 9164626052791124 even though the last timestamp was 9164626053263961. Stopping the eBPF tracker.
<probe> INFO: 2017/03/21 12:20:51.430468 Publish loop for cloud.weave.works starting
<probe> INFO: 2017/03/21 12:20:51.430899 Publish loop for scope.kube-system.svc.cluster.local starting
<probe> INFO: 2017/03/21 12:20:51.431330 Publish loop for frontend.dev.weave.works starting
<probe> ERRO: 2017/03/21 12:20:51.604695 tcp tracer received event with timestamp 9164627054285071 even though the last timestamp was 9164627055437225. Stopping the eBPF tracker.
<probe> ERRO: 2017/03/21 12:20:52.992805 tcp tracer received event with timestamp 9164628053474331 even though the last timestamp was 9164628056281821. Stopping the eBPF tracker.
<probe> ERRO: 2017/03/21 12:20:53.004402 tcp tracer received event with timestamp 9164628054192927 even though the last timestamp was 9164628054916996. Stopping the eBPF tracker.
<probe> ERRO: 2017/03/21 12:20:53.059531 tcp tracer received event with timestamp 9164629052464334 even though the last timestamp was 9164629055395006. Stopping the eBPF tracker.
<probe> ERRO: 2017/03/21 12:20:53.106566 tcp tracer received event with timestamp 9164629052790759 even though the last timestamp was 9164629055350544. Stopping the eBPF tracker.
<probe> ERRO: 2017/03/21 12:20:53.262504 tcp tracer received event with timestamp 9164629053085192 even though the last timestamp was 9164629053624826. Stopping the eBPF tracker.
<probe> ERRO: 2017/03/21 12:20:53.582277 tcp tracer received event with timestamp 9164629054868485 even though the last timestamp was 9164629055992184. Stopping the eBPF tracker.
<probe> ERRO: 2017/03/21 12:20:53.594538 tcp tracer received event with timestamp 9164629052317434 even though the last timestamp was 9164629055942911. Stopping the eBPF tracker.
<probe> ERRO: 2017/03/21 12:20:54.192466 tcp tracer received event with timestamp 9164630053260029 even though the last timestamp was 9164630054782341. Stopping the eBPF tracker.
<probe> ERRO: 2017/03/21 12:20:54.265299 tcp tracer received event with timestamp 9164630054651380 even though the last timestamp was 9164630055673123. Stopping the eBPF tracker.
<probe> ERRO: 2017/03/21 12:20:54.325498 tcp tracer received event with timestamp 9164630054710055 even though the last timestamp was 9164630054996914. Stopping the eBPF tracker.
<probe> ERRO: 2017/03/21 12:20:54.385495 tcp tracer received event with timestamp 9164630053596825 even though the last timestamp was 9164630055770522. Stopping the eBPF tracker.
<probe> ERRO: 2017/03/21 12:20:54.479220 tcp tracer received event with timestamp 9164630052901363 even though the last timestamp was 9164630054585827. Stopping the eBPF tracker.
<probe> ERRO: 2017/03/21 12:20:54.482689 tcp tracer received event with timestamp 9164630052403622 even though the last timestamp was 9164630052901363. Stopping the eBPF tracker.
<probe> ERRO: 2017/03/21 12:20:54.675139 tcp tracer received event with timestamp 9164630052662181 even though the last timestamp was 9164630055610166. Stopping the eBPF tracker.
```
